File: 2020/day15/day15.jl
function solve(line::Vector{Int}, iterations::Int)
    N = length(line)
    spoken_time = Dict(line[i] => i for i in 1:N-1 )
    last_number = line[end]

    for i in N:iterations-1
        temp_num = if last_number in keys(spoken_time)
            i - spoken_time[last_number]
        else
            0
        end
        spoken_time[last_number] = i
        last_number = temp_num
    end

    return last_number
end

function main(filename::String)
    lines = readlines(filename) |>
        x->split.(x, ',') |> x->map.(f->parse(Int64,f), x)

    for line in lines
        println("Part1: ", solve(line, 2020))
        println("Part2: ", solve(line, 30000000))
    end


end

File: 2020/day15/day15.cpp
#include "../../utils.hpp"
#include <algorithm>
#include <unordered_map>

using namespace std;

auto solve3(vector<i64> &line, int iterations) {
  vector<int> spoken_time;
  spoken_time.reserve(iterations);
  const auto N = line.size();
  for (int i = 0; i < N - 1; ++i) {
    spoken_time.push_back(line[i]);
  }

  int next = line.back();
  for (int i = N - 1; i < iterations - 1; ++i) {
    auto ptr = find(spoken_time.begin(), spoken_time.end(), next);
    int tmp =
        ptr != spoken_time.end() ? i - distance(spoken_time.begin(), ptr) : 0;
    spoken_time[i] = next;
    next = tmp;
  }
  return next;
}
auto solve2(vector<i64> &line, int iterations) {
  const auto N = line.size();
  unordered_map<int, int> spoken_time;
  for (int i = 0; i < N - 1; ++i) {
    spoken_time[line[i]] = i;
  }

  int next = line.back();
  for (int i = N - 1; i < iterations - 1; ++i) {
    auto ptr = spoken_time.find(next);
    int tmp = ptr != spoken_time.end() ? i - ptr->second : 0;
    spoken_time.insert_or_assign(next, i);
    next = tmp;
  }
  return next;
}

int main(int arc, char *argv[]) {
  auto lines = utils::map(utils::to_int, utils::split_by(utils::read_file_as_string_view(argv[1]), ','));

  // cout << solve(lines, 2020) << '\n';
  cout << solve2(lines, 2020) << '\n';
  // cout << solve3(lines, 2020) << '\n';
  cout << solve2(lines, 30000000) << '\n';
  return 0;
}
